frank/utilities.py

```diff
@@ -272,7 +321,77 @@ def cut_data_by_baseline(u, v, vis, weights, cut_range):
 
     return u_cut, v_cut, vis_cut, weights_cut
 
+def estimate_weights(u, v, V, nbins=300, log=True, use_median=False):
+    r"""
+    Estimate the weights using the variance of the binned visibilities. 
+
+    The estimation is done assuming that the variation in each bin is dominated
+    by the noise. This will be true if:
+        1) The source is axi-symmetric,
+        2) The uv-points have been deprojected,
+    Otherwise the variance may be dominated by real variations in the 
+    visibilities.
+
+    Parameters
+    ----------
+    u, v : array, unit = :math:`\lambda`
+        u and v coordinates of observations (deprojected).
+    V : array, unit = Jy
+        Obsersed visibility. If complex, the weights will be computed from the
+        average of the variance of the real and imaginary components, as in
+        CASA's statwt. Otherwise the variance of the real part is used.
+    nbins : int, default=300
+        Number of bins used.
+    log : bool, default=True
+        If True, the uv bins will be constructed in log space, otherwise linear
+        spaced bins will be used.
+    use_median : bool, default=False
+        If True all of the weights will be set to the median of the variance
+        estimated across the bins. Otherwise, the baseline dependent variance 
+        will be used.
+
+    Returns
+    -------
+    weights : array,
+        Esimtate weight for each uv point.
+
+    Notes
+    -----
+        - This function does not support datasets where the weights may be
+          expected to be different for different baselines, e.g. if the data
+          includes visibilities from telescopes of different sizes
+        - Bins with only one uv point will not have a variance estimate. The
+          median across all baselines is used instead.
+    """
+    logging.info('  Estimating visibility weights.')
+
+    q = np.hypot(u,v)
+    if log:
+        q = np.log(q)
+
+    q -= q.min()
```

Comment:
I don't follow why you need to do this instead of leaving `q` as the absolute baselines?

Reply:
Several reasons: 

1. For log, q.min() could be negative
2. The nbins are more effectively used spreading them from q.min() to q.max() than 0 to q.max(), especially for log.

Reply:
`q.min()` could be negative for log if `q.min` is < 1 yes, but you'll never have a baseline at 1 \lambda for mm observations! 

Couldn't you just do `bin_width = (q.max() - q.min()) / nbins` in l.374?

I know this is all internal to `estimate_weights`, so if you're adamant it needs to be done, ok, but I just think it's better practice / more foolproof to keep the baselines in absolute units.

Reply:
You still need to make sure that all q > 0

Reply:
I could only do it for the log case

Reply:
I must be missing something, when will q ever be negative? u and v can individually be negative, but q is always positive, it's a physical distance.

Reply:
log(q) < 0 if q < 1

Reply:
Right, which is why I noted q = 1 is a baseline at 1 \lambda, which you'll never have (can't separate antennas by ~1 mm).

Reply:
I've changed the code now. The original worked more like np.histogram with bins=nbins though.

Reply:
It still makes sense to subtract qmin in the log case because loq(qmin) can be ~ 4-5 while log(qmax) - log(qmin) ~ 1-3. You're just creating a bunch of empty bins for no reason.

Reply:
ok yes that's true, I see. Thanks for updating where possible.